### ISSUE TITLE
Failed Commits-Squash Corrupts Repository (SM 2000)

### ISSUE THREAD

**Issue by tajmone** [sublimehq/sublime_merge#687]
**Version info**

 - OS:  Windows 10 x64
 - Build: 2000

**Description**

When I try squash together a group of commits, and the operation fails, I'm left with a corrupt repository. I'm not sure what the exact problem is, but it seems to boil down to the following message found in the `git status` message:

```
$ git status
On branch temp-squash
git-rebase-todo is missing.
No commands done.
No commands remaining.
You are currently editing a commit while rebasing branch 'temp-squash' on '(NULL)'.
  (use "git commit --amend" to amend the current commit)
  (use "git rebase --continue" once you are satisfied with your changes)

nothing to commit, working tree clean
```

The missing `git-rebase-todo` prevents the `git rebase --abort` command:

    error: could not read '.git/rebase-merge/onto': No such file or directory

The only way out I could fin was to use:

    rm -rf .git/rebase-merge

**Steps to reproduce**

I'm not sure, it happened when I select a number of commits and then right-click on "Edit Commit » Squash Selected Commits" and then the operation fails (for whatever reason).

**Comment by dpjohnst** [comment]
Hi @tajmone,

Thank you for reporting this issue. To help diagnose the issue, could you provide me with the version of Git that Sublime Merge is using?

You can find this information by opening Sublime Merge, and selecting Help > Debug Information.

**Comment by tajmone** [comment]
> o help diagnose the issue, could you provide me with the version of Git that Sublime Merge is using?

    git version 2.19.1.windows.1

**Comment by dpjohnst** [comment]
Hi @tajmone,

I have so far been unable to reproduce this issue on my end. It seems as if Sublime Merge is failing to write the required rebase files to the git directory.

Does this happen consistently for you when you squash a group of commits, or intermittently?

Thanks,
\- Dylan

**Comment by tajmone** [comment]
> Does this happen consistently for you when you squash a group of commits, or intermittently?

Consistently. I had reattempted to squash the above mentioned commits many times overs, hoping that it would ultimately work, but the problem kept showing up in SM, so I ultimately did it via Git CLI (and haven't attempted to use SM for this since then).

**Comment by tajmone** [comment]
If this might help, the repository I was working on is this one:

https://github.com/tajmone/Tonton-Pixel-Photoshop-Scripts

I was attempting to squash togeter multiple commits to make the history cleaner, and had to do it via Git CLI due to SM failing.

Since I've overwritten the original history (from 38 commits down to 3), the original repository only survives as a fork, by someone else:

https://github.com/won21kr/Tonton-Pixel-Photoshop-Scripts

Basically, I've left untouched the last 2 commits and squashed together all the initial 36 commits.

This is the repository that displayed the SM squash problem. I hadn't tried it with other repos, so I'm not sure if its a problem linked to that repository, but since interactive rebase via Git CLI worked fine at first attempt, I tend to think that it's a SM related problem.

**Comment by dpjohnst** [comment]
Many thanks for that information, and thank you for your patients while I investigated this issue. 

I was able to reproduce the error, and I am in the process of creating a fix. I'll keep you updated, but you can expect this issue to be fixed in the next build of Sublime Merge.

Cheers,
\- Dylan

**Comment by dpjohnst** [comment]
Hi @tajmone,

Could you confirm whether the problem is fixed for you in Build 2003?

Thanks,
\-Dylan

**Comment by tajmone** [comment]
> Could you confirm whether the problem is fixed for you in Build 2003?

I've just updated SM to Dev Build 2004 (on the Dev channel there's no 2003 build, it bumped from 2002 directly to 2004).

I can confirm that the repository is not being corrupted any more by failed commit-squash operations in SM! But it seems to me that some operations are not being executed either ...

I've kept a local copy of the original `master` branch from the above mentioned repository, as it was before the commit squash operations, so I' trying to replay the same operations that led to the original problem of this Issue. So I'll be referring here to commit hashes as found on

https://github.com/won21kr/Tonton-Pixel-Photoshop-Scripts

For example, attempting to use <kbd>Edit Commit</kbd> » <kbd>Squash with parent</kbd> on [4579bcd] silently fails — whereas usually squash operations either ask for confirmation or result in an error warning box.

I've tried various commit squash operations (squash with parent/selected commits) on the original branch, and noticed that some ask for confimation and succeed, while others fail with the following error (without corrupting the repository):

> Error: only commits reachable from head can be modified

... then there also those squash operations that simply do nothing and don't report anything either (as mentioned above).

It's worth mentioning here that I've reported a bug about SM UI not always refreshing after commit squash operations, so sometimes it just looks like nothing happened because the sidebar with the commits it's not being updated (but in the above example I check via Git CLI and can confirm that nothing was changed in the repository after attempting the commits squash operation).

Unfortunately, I don't understand deeply enough how Git is supposed to handle commit squash operations, so I'm unable to evaluate wether SM is behaving as expected in various contexts. So I'm unable to confirm wether everything is working fine, beside the fact that the repository is not corrupting any more.

Also, since my local repository contains divergent histories there might be additional problems at play here — I did experience problems in the past with SM handling repositories with orphan branches.


[4579bcd]: https://github.com/won21kr/Tonton-Pixel-Photoshop-Scripts/commit/4579bcdeb2bceba5b172ed22dab394ec22ac66d7

**Comment by tajmone** [comment]
@dpjohnst, I've pushed my local copy of the original `master` branch on the repository:

https://github.com/tajmone/Tonton-Pixel-Photoshop-Scripts/tree/master-old

So you can clone it and try it locally.

During these tests I've noticed some other bugs (UI-related) which I'm about to report.

**Comment by tajmone** [comment]
My general impression (as a long-time SM Win user) is that SM becomes erratic when a repository contains orphan branches. For this reason, I usually create a separate repository for a GitHub Pages orhpan branch, instead of packing it with main repo.

So, my guess is that SM fails to carry out diveregency operations that could lead to an orphan branch, and that if such divergency is introduced via Git CLI, then SM starts to fail some common commands on all branches — as if the presence of orphan branches "confuses" SM internal engine.

The above linked repository is a good test ground to check for erratic behavior.

**Comment by dpjohnst** [comment]
Hi @tajmone,

Thank you for the feedback you've given - all has been recorded on my end, and I will be following up on it. In the meantime, I'm going to close this issue as the original issue has been fixed.

I'll follow up on the other issues as more progress is made :)

Kind regards,
\- Dylan
